pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go

```diff
@@ -0,0 +1,144 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+	"sigs.k8s.io/mcs-api/pkg/client/clientset/versioned"
+
+	"istio.io/istio/pilot/pkg/model"
+	serviceRegistryKube "istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/queue"
+	"istio.io/pkg/log"
+)
+
+type ServiceExportController struct {
+	ServiceExportOptions
+	client        versioned.Interface
+	serviceClient corev1.CoreV1Interface
+
+	queue           queue.Instance
+	serviceInformer cache.SharedInformer
+
+	// We use this flag to short-circuit the logic and stop the controller
+	// if the CRD does not exist (or is deleted)
+	mcsSupported bool
+}
+
+// ServiceExportOptions provide options for creating a ServiceExportController.
+type ServiceExportOptions struct {
+	Client       kube.Client
+	ClusterID    string
+	DomainSuffix string
+	ClusterLocal model.ClusterLocalProvider
+}
+
+// NewServiceExportController creates a new ServiceExportController.
+func NewServiceExportController(opts ServiceExportOptions) *ServiceExportController {
+	c := &ServiceExportController{
+		ServiceExportOptions: opts,
+		client:               opts.Client.MCSApis(),
+		serviceClient:        opts.Client.Kube().CoreV1(),
+		queue:                queue.NewQueue(time.Second),
+		mcsSupported:         true,
+	}
+
+	c.serviceInformer = opts.Client.KubeInformer().Core().V1().Services().Informer()
+	c.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) { c.onServiceAdd(obj) },
+		// Do nothing on delete. When we create ServiceExport, we bind its
```

Comment:
I would also comment: "Do nothing on update. The controller only acts on parts of the Service that are immutable" or similar